docs/source/llms/notebooks/chat-model-tool-calling.ipynb

```diff
@@ -0,0 +1,467 @@
+     "text": [
+      "/Users/daniel.lok/miniconda3/envs/test/lib/python3.9/site-packages/mlflow/models/utils.py:523: FutureWarning: Since MLflow 2.16.0, we no longer convert dictionary input example to pandas Dataframe, and directly save it as a json object. If the model expects a pandas DataFrame input instead, please pass the pandas DataFrame as input example directly.\n",
+      "  example = _Example(input_example)\n",
+      "Downloading artifacts: 100%|██████████| 7/7 [00:00<00:00, 216.81it/s]\n"
+     ]
+    },
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Received a tool call, calling the weather tool...\n"
+     ]
+    },
+    {
+     "name": "stderr",
+     "output_type": "stream",
+     "text": [
+      "2024/10/25 11:39:29 INFO mlflow.tracking._tracking_service.client: 🏃 View run gaudy-rook-875 at: http://localhost:5000/#/experiments/0/runs/567216afacbe4cd9bf2fcbe019e71b5b.\n",
+      "2024/10/25 11:39:29 INFO mlflow.tracking._tracking_service.client: 🧪 View experiment at: http://localhost:5000/#/experiments/0.\n"
+     ]
+    }
+   ],
+   "source": [
+    "# messages to use as input examples\n",
+    "messages = [\n",
+    "    {\"role\": \"system\", \"content\": \"Please use the provided tools to answer user queries.\"},\n",
+    "    {\"role\": \"user\", \"content\": \"What's the weather in Singapore?\"},\n",
+    "]\n",
+    "\n",
+    "input_example = {\n",
+    "    \"messages\": messages,\n",
+    "}\n",
+    "\n",
+    "# instantiate the model\n",
+    "model = WeatherModel()\n",
+    "\n",
+    "# log the model\n",
+    "with mlflow.start_run():\n",
+    "    model_info = mlflow.pyfunc.log_model(\n",
+    "        artifact_path=\"weather-model\",\n",
+    "        python_model=model,\n",
+    "        input_example=input_example,\n",
+    "    )"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Using the model for generations\n",
+    "\n",
+    "Now that the model is logged, our work is more or less done! In order to use the model for predictions, let's load it back using `mlflow.pyfunc.load_model()`.\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 25,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stderr",
+     "output_type": "stream",
+     "text": [
+      "Downloading artifacts: 100%|██████████| 7/7 [00:00<00:00, 160.13it/s]\n"
+     ]
+    },
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Received a tool call, calling the weather tool...\n",
+      "The weather in Singapore is sunny, with a temperature of 20°C.\n",
+      "Received a tool call, calling the weather tool...\n",
+      "The weather in San Francisco is sunny, with a temperature of 20°C.\n"
+     ]
+    }
+   ],
+   "source": [
+    "import mlflow\n",
+    "\n",
+    "# Load the previously logged ChatModel\n",
+    "tool_model = mlflow.pyfunc.load_model(model_info.model_uri)\n",
```

Comment:
Let's print out the model URI because we need it later when running `mlflow models serve -m`.

Reply:
done!